plugins/state_history_plugin/state_history_plugin.cpp

```diff
@@ -128,44 +189,28 @@ struct state_history_plugin_impl : std::enable_shared_from_this<state_history_pl
                    fc::datastream<const char*> ds(d, s);
                    state_request               req;
                    fc::raw::unpack(ds, req);
-                   std::visit(*self, req);
+                   std::visit(*self, std::move(req));
                    self->start_read();
                 });
              });
       }
 
-      void send(const char* s) {
-         send_queue.push_back({s, s + strlen(s)});
-         send();
-      }
-
       template <typename T>
-      void send(T obj) {
-         send_queue.push_back(fc::raw::pack(state_result{std::move(obj)}));
-         send();
-      }
-
-      void send() {
-         if (sending)
-            return;
-         if (send_queue.empty())
-            return send_update();
-         sending = true;
-         derived_session().socket_stream->binary(sent_abi);
-         sent_abi = true;
-         derived_session().socket_stream->async_write( //
-             boost::asio::buffer(send_queue[0]),
-             [self = derived_session().shared_from_this()](boost::system::error_code ec, size_t) {
-                self->callback(ec, "async_write", [self] {
-                   self->send_queue.erase(self->send_queue.begin());
-                   self->sending = false;
-                   self->send();
-                });
-             });
+      void send(T obj, ::std::optional<::fc::zipkin_span>&& span) {
+         if (!send_thread_has_exception) {
```

Comment:
Seems like we should at least log that we are ignoring this send.

Reply:
The exception is reported via post at the end of thread function.

Reply:
Seems racy. `send_thread_has_exception` won't be set true until some period of time after the io_context has stopped running. So it seems possible for `send()` to queue up another `post()` that will never be dispatched. If that's not a problem (and it might not be, I'm not sure on cursory look), then I don't see why this check needs to be here in the first place.

Reply:
The loop in the thread will break when the first task throw an exception, all the subsequent tasks would not be able to be executed. The check here is not strictly needed, just intended to be save some cycles.